### ISSUE TITLE
Issue in Routing and Squash

### ISSUE THREAD

**Issue by balassbals** [cedrickchee/capsule-net-pytorch#3]
In routing method, F.softmax is called on b_ij which is 1 x 1152 x 10 x 1 but it should be called on dimension 2. Of course pytorch 0.2.0 does not yet have softmax where dim can be supplied  but using F.softmax as done here may not behave as required.

Also the default value for dim param in squash function is 2 but in line 93 when it is called on s_j to compute v_j, default dimension will not work since s_j is 128 x 1 x 10 x 16 x 1 and so dim = 3 should be passed. 
Correct me if i'm wrong. But if I'm correct and with these changes can we get in 10 epochs the expected results?

**Comment by cedrickchee** [comment]
Hi. Thank you for reporting the issue. I will check this and get back to you soon.

**Comment by cedrickchee** [comment]
**Problem 1:**
> In routing method, F.softmax is called on b_ij which is 1 x 1152 x 10 x 1 but it should be called on dimension 2.

Yes, you are right. Good catch! I have fixed this bug.

> Of course pytorch 0.2.0 does not yet have softmax where dim can be supplied but using F.softmax as done here may not behave as required.

When PyTorch 0.4.0 stable lands, it will support softmax function that take a dimension. In the meantime, I have implemented a custom softmax function to resolve this issue.

**Problem 2:**
> Also the default value for dim param in squash function is 2 but in line 93 when it is called on s_j to compute v_j, default dimension will not work since s_j is 128 x 1 x 10 x 16 x 1 and so dim = 3 should be passed.

The squash function was implemented based on naturomics's CapsNet-Tensorflow:

`vec_squared_norm = tf.reduce_sum(tf.square(vector), -2, keep_dims=True)`

In naturomics's implementation, the dim = -2. I couldn't wrap my head around why you think we should pass dim = 3 to the sum ops. Could you please enlighten me?

> Correct me if i'm wrong. But if I'm correct and with these changes can we get in 10 epochs the expected results?

I can't confirm this as of now until after we understand whether Problem 2 affect the results.

**Comment by balassbals** [comment]
As per the paper the capsule's length need to be squashed. v_j's length in line 93 is 16 which is in dim 3 which is same as -2. But in your squash function dim is fixed to be 2. 

Just want to know if you could get the results after these changes  because I could not get.

**Comment by cedrickchee** [comment]
> As per the paper the capsule's length need to be squashed. v_j's length in line 93 is 16 which is in dim 3 which is same as -2. But in your squash function dim is fixed to be 2.

Fixed.

> Just want to know if you could get the results after these changes because I could not get.

What is the results you are getting? Can you share it here?

Well, finally, I think we should be able to reproduce the results in the paper after these fixes and improvements I have made in the latest commits, especially this commit 66b65e044c19680cb76a31dad6cc983b4aed6375.

With these changes, I have re-train the model. The training is still in progress and I think the results looks promising. Here are some early results of the training and testing loss and accuracy of CapsNet-v4 at epoch 11 of 50:

`python main.py -epochs 50 --batch-size 128 --lr 0.01 --num-routing 3 --use-reconstruction-loss yes --regularization-scale 0.0005`

- Test loss: 0.2293

![test_losses](https://user-images.githubusercontent.com/145605/33239381-2d09dcb8-d2dc-11e7-839a-a3e34b2ef88e.png)

- Test accuracy: 99.07%

![test_accuracy](https://user-images.githubusercontent.com/145605/33239397-78f27d38-d2dc-11e7-82f7-513d40f4adc5.png)

- Train loss: 0.2110

![train_total_loss](https://user-images.githubusercontent.com/145605/33239410-b4349caa-d2dc-11e7-95c8-8a46e1356b0c.png)

- Train accuracy: 99.60%

![train_accuracy](https://user-images.githubusercontent.com/145605/33239414-bdcbac90-d2dc-11e7-8917-3d9b9477dcb4.png)

**Comment by balassbals** [comment]
Thanks for the updates. Currently busy preparing to offer a tutorial on CNN from Dec 7-9. I will get back to you may be before that if i get my code working or after that. Actually the loss for me is not going down below 1.82. I trained for 100 epochs. May be somewhere else I'm missing something. Will get back.

**Comment by cedrickchee** [comment]
Thank you for the updates. I see. No rush. Just don't worry about it. I am busy as well. In the meantime, I will try to see if I can replicate your case when I am more available.

**Comment by balassbals** [comment]
One more issue. I think in line num164 (**return utils.squash(unit, dim=2)**) in capsule_layer.py, dim to be passed should be 1 and not 2. We should squash vectors of length 8.

**Comment by balassbals** [comment]
By the way I brought down the error rate to 0.93 in my implementation (by removing ReLU in the first convolution before primary capsule) after some 90 epochs but the accuracy was mere 9%. So I took a chance and squashed in the primary capsule along the dimension of 1152 capsules instead of vector length of 8 and the test accuracy shoots up to 97% in just 10 epochs. I understand from the paper that we should squash along vector dimension but the other way works.

**Comment by balassbals** [comment]
Got my implementation working. Test accuracy is 99.37% Squashing done along vector dimensions (8 and 16) as mentioned in paper.

**Comment by cedrickchee** [comment]
I am sorry for the late reply. I was swarmed with work for the past month or so. I really appreciate your help in testing and reverting results in the early days not long after opened up this repo to the public.

It's good to hear that you managed to get your implementation working. Thank you for pointing a few things in the paper that I think I might missed or interpret incorrectly and thus affecting the logic in the implementation.

Are you OK if I close this issue? Thank you.
